[Feature][Quantization] MXFP4 support for MOE models

fxmarty-amd opened the pull request:
This PR follows https://github.com/vllm-project/vllm/pull/16943, and adds the possibility to load MOE models using MXFP4 weights with dynamic per-group MXFP4 quantization for activations.

We did not yet release such models publicly, but expect to release some soon.

At the moment, execution on MI300 runs a simulated scheme where weights are dequantized on the fly, and QDQ is done on activations on the fly, using HIP kernels 

Left to do:
- [x] Add test.
- [x] Add documentation.
- [ ] Implement the code path for real mxfp4 * mxfp4 GEMM (maybe in an other PR)
- [x] Validate sensible eval results for Deepseek R1, llama 4 and llama 405B

fxmarty-amd committed:
MXFP4

wip

wip & debug

update

cleanup

use quark realquantizer for pack/quant/dequant

comment on cudagraph issue; remove prints

Keep only 1 place importing quark

cudagraph issue resolved; dq weight at load time for efficiency

Signed-off-by: Bowen Bao <bowenbao@amd.com>

lint

Signed-off-by: Bowen Bao <bowenbao@amd.com>

turn on emulation based on platform

Signed-off-by: Bowen Bao <bowenbao@amd.com>

add fused moe support - ugly wip

running version

Add envar if dequant weight at load time

Signed-off-by: Bowen Bao <bowenbao@amd.com>

Mxfp4 memory leak fixes (#2)

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
wip

wip & debug

update

cleanup

use quark realquantizer for pack/quant/dequant

comment on cudagraph issue; remove prints

Keep only 1 place importing quark

cudagraph issue resolved; dq weight at load time for efficiency

Signed-off-by: Bowen Bao <bowenbao@amd.com>

lint

Signed-off-by: Bowen Bao <bowenbao@amd.com>

turn on emulation based on platform

Signed-off-by: Bowen Bao <bowenbao@amd.com>

add fused moe support - ugly wip

running version

Add envar if dequant weight at load time

Signed-off-by: Bowen Bao <bowenbao@amd.com>

Mxfp4 memory leak fixes (#2)

Fix VLLM_QUARK_EMU_MEM_OPT route

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
wip

wip & debug

update

cleanup

use quark realquantizer for pack/quant/dequant

comment on cudagraph issue; remove prints

Keep only 1 place importing quark

cudagraph issue resolved; dq weight at load time for efficiency

Signed-off-by: Bowen Bao <bowenbao@amd.com>

lint

Signed-off-by: Bowen Bao <bowenbao@amd.com>

turn on emulation based on platform

Signed-off-by: Bowen Bao <bowenbao@amd.com>

add fused moe support - ugly wip

running version

Add envar if dequant weight at load time

Signed-off-by: Bowen Bao <bowenbao@amd.com>

Mxfp4 memory leak fixes (#2)

Fix VLLM_QUARK_EMU_MEM_OPT route

Signed-off-by: Felix Marty <felmarty@amd.com>

BowenBao committed:
Separate moe to another PR

Signed-off-by: Bowen Bao <bowenbao@amd.com>

BowenBao committed:
lint

Signed-off-by: Bowen Bao <bowenbao@amd.com>

BowenBao committed:
large moe support

Signed-off-by: Felix Marty <felmarty@amd.com>

BowenBao committed:
large moe support

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
use kernels

Signed-off-by: Felix Marty <felmarty@amd.com>

BowenBao committed:
use dynamic quant kernel for moe activation

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
add kernel/non-kernel branches for mxfp4

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
set VLLM_QUARK_MXFP4_Q_DQ_QDQ_IMPLEM to 'hip', 'triton' or 'torch' to select the q/dq/qdq implem for mxfp4

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
fix

Signed-off-by: Felix Marty <felmarty@amd.com>

BowenBao committed:
Move all kernels into Quark (#3)

Co-authored-by: Felix Marty <felmarty@amd.com>
Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
rebase fixup

Signed-off-by: Felix Marty <felmarty@amd.com>

DarkLight1337 commented:
Can you merge from main to fix pre-commit?

fxmarty-amd committed:
Merge branch 'main' into mxfp4_moe

fxmarty-amd committed:
style

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
fix style

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
add test and documentation

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
style

Signed-off-by: Felix Marty <felmarty@amd.com>

BowenBao commented:
@mgoin ptal when you get a chance, thanks!

fxmarty-amd committed:
Merge branch 'main' into mxfp4_moe

fxmarty-amd committed:
fix conflicts

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
style

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
style bis

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
Merge branch 'main' into mxfp4_moe

fxmarty-amd committed:
add accuracy test

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
style

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd commented:
@DarkLight1337 @mgoin WDYT? We expect to add a CDNA4 native execution path running GEMMs in mxfp4 in a next PR.

fxmarty-amd committed:
fix test

Signed-off-by: Felix Marty <felmarty@amd.com>

mgoin reviewed (commented):
Thanks for refactoring the shared qdq utils and adding model smoke tests. This seems reasonable to me, just a few nits

fxmarty-amd committed:
address review comments

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd commented on tests/quantization/test_quark.py:
```diff
@@ -89,3 +95,83 @@ def test_quark_fp8_parity(vllm_runner):
+    tp: int
+
+
+@pytest.mark.parametrize('model_case', [
+    ModelCase("fxmarty/qwen_1.5-moe-a2.7b-mxfp4", tp=1),
+    ModelCase("fxmarty/deepseek_r1_3_layers_mxfp4", tp=8),
+    ModelCase("fxmarty/Llama-4-Scout-17B-16E-Instruct-2-layers-mxfp4", tp=1)
+])
```
@mgoin good catch, it seems `load_format="dummy"` works well.

I am not very familiar with vllm CI - it seems https://github.com/vllm-project/vllm/blob/1b1e8e05ff3c26b98e4161bd3c8671e86fb145f4/.buildkite/test-pipeline.yaml#L410-L417 should trigger these tests on `amdexperimental`, but it seems not to be in https://buildkite.com/vllm/fastcheck/builds/24374#0196e8fe-ef69-42b2-ade4-4a6dbbaa11c4 in the current branch.

It seems to be run on `main`, though, e.g. https://buildkite.com/vllm/ci/builds/20293#0196e18d-6ec4-497d-b617-d2b590af9f7c

The thing is that they rely on the next amd-quark release - this is addressed in https://github.com/vllm-project/vllm/pull/17888/commits/f5e5fbb027fb81155f9c58e9e6c522b7e7827fcf.

fxmarty-amd committed:
Merge branch 'main' into mxfp4_moe

fxmarty-amd committed:
merge fixes

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
style

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
skip tests if not enough gpus

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
typo

Signed-off-by: Felix Marty <felmarty@amd.com>

fxmarty-amd committed:
Merge branch 'main' into mxfp4_moe

fxmarty-amd committed:
Merge branch 'main' into mxfp4_moe

fxmarty-amd commented:
Hi @mgoin, I reran `pytest tests/quantization/test_quark.py -s -vvvvv` on this branch with main merged today, it looks good. Let me know if this PR needs any more updates :)

AMD is gently starting to release OCP MXFP4 models publicly: https://huggingface.co/collections/amd/quark-quantized-mxfp4-models-68068f8c965d9267a996616d

mgoin commented:
Thanks for the ping, I'll take a look

mgoin commented on vllm/model_executor/layers/quantization/utils/mxfp4_utils.py:
```diff
@@ -1,45 +1,29 @@
-        quant_dtype="fp4",
-    )
+def quant_dequant_mxfp4(x: torch.Tensor,
+                        scale_calculation_mode: str = "even") -> torch.Tensor:
+    try:
+        from quark.torch.kernel import mx
+    except ImportError as err:
+        raise ImportError("The package `amd-quark` is required to use "
+                          "MX-FP4 models. Please install it with `pip install "
+                          "amd-quark`.") from err
 
-    return x, scale
+    return mx.qdq_mxfp4(x, scale_calculation_mode)
```
It would be nice to have at least a naive torch implementation to compare basic correctness against this external kernel. Usually we would upstream a kernel like this for a core quantization like mxfp4

fxmarty-amd replied:
Added an (ugly) pure torch reference in `tests/quantization` in https://github.com/vllm-project/vllm/pull/17888/commits/42d97884fbdf13d348414008d69f7c423601f3af

mgoin commented on vllm/model_executor/layers/fused_moe/fused_moe.py:
```diff
@@ -982,15 +987,16 @@ def get_config_dtype_str(
         return torch.float8_e4m3fn
     elif use_int8_w8a8:
         return torch.int8
+    elif use_mxfp4_w4a4:
+        # TODO: replace by a torch.dtype once PyTorch natively supports mxfp4.
+        return "mxfp4"
```
How is this qtype used? It seems potentially dangerous to not use a torch.dtype here

mgoin commented on vllm/model_executor/layers/fused_moe/fused_moe.py:
```diff
@@ -1235,6 +1248,15 @@ def fused_experts_impl(
         assert hidden_states.shape[1] // 2 == w1.shape[
             2], "Hidden size mismatch"
+    elif use_mxfp4_w4a4:
+        if current_platform.supports_mx() or envs.VLLM_QUARK_EMU_MEM_OPT:
+            # 16bit activation and fp4x2 packed weight
+            assert hidden_states.shape[1] // 2 == w1.shape[
+                2], "hidden size mismatch"
+        else:
+            # 16bit activation and dequantized 16bit weight
+            assert hidden_states.shape[1] == w1.shape[
+                2], "hidden size mismatch"
```
I know we discussed this in the dense MXFP PR but I think we should eventually remove the `not envs.VLLM_QUARK_EMU_MEM_OPT` case. Of course it will be slower for inference, but the memory benefit of the lower precision is the expected UX of this format and this complexity is not nice to keep propagating.

fxmarty-amd committed:
add missing args in examples

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
remove VLLM_QUARK_EMU_MEM_OPT, always keeps mxfp4 weights in low precision

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
use emulate=True for mxfp4 gemm on cdna4 until real kernels are integrated

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
add slow/non-optimized reference torch mxfp4 quant and qdq implementation

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
style

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
style 2

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd commented:
I think the CI looks good, apart from some bitsandbytes tests that may be unrelated:
```
[2025-06-17T16:53:16Z] [31mFAILED[0m quantization/test_bitsandbytes.py::[1mtest_load_4bit_bnb_model[facebook/opt-125m-quantize opt model inflight][0m - AssertionError: function <function test_load_4bit_bnb_model at 0x7f4a4e55ec00> failed when called with args () and kwargs {'hf_runner': <class 'tests.conftest.HfRunner'>, 'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'example_prompts': ['vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n', 'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\n', 'Compare and contrast artificial intelligence with human intelligence in terms of processing information.\n', 'Describe the basic components of a neural network and how it can be trained.\n', 'Write a short story about a robot that dreams for the first time.\n', 'Analyze the impact of the COVID-19 pandemic on global economic structures and future business models.\n', 'Explain the cultural significance of the Mona Lisa painting, and how its perception might vary in Western versus Eastern societies.\n', "Translate the following English sentence into Japanese, French, and Swahili: 'The early bird catches the worm.'\n"], 'model_name': 'facebook/opt-125m', 'description': 'quantize opt model inflight'}
[2025-06-17T16:53:16Z] [31mFAILED[0m quantization/test_bitsandbytes.py::[1mtest_load_4bit_bnb_model[mistralai/Mistral-7B-Instruct-v0.3-quantize inflight model with both HF and Mistral format weights][0m - AssertionError: function <function test_load_4bit_bnb_model at 0x7f4a4e55ec00> failed when called with args () and kwargs {'hf_runner': <class 'tests.conftest.HfRunner'>, 'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'example_prompts': ['vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n', 'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\n', 'Compare and contrast artificial intelligence with human intelligence in terms of processing information.\n', 'Describe the basic components of a neural network and how it can be trained.\n', 'Write a short story about a robot that dreams for the first time.\n', 'Analyze the impact of the COVID-19 pandemic on global economic structures and future business models.\n', 'Explain the cultural significance of the Mona Lisa painting, and how its perception might vary in Western versus Eastern societies.\n', "Translate the following English sentence into Japanese, French, and Swahili: 'The early bird catches the worm.'\n"], 'model_name': 'mistralai/Mistral-7B-Instruct-v0.3', 'description': 'quantize inflight model with both HF and Mistral format weights'}
[2025-06-17T16:53:16Z] [31mFAILED[0m quantization/test_bitsandbytes.py::[1mtest_load_pre_quant_4bit_bnb_model[PrunaAI/Einstein-v6.1-Llama3-8B-bnb-4bit-smashed-read pre-quantized 4-bit FP4 model][0m - AssertionError: function <function test_load_pre_quant_4bit_bnb_model at 0x7f4a42f18e00> failed when called with args () and kwargs {'hf_runner': <class 'tests.conftest.HfRunner'>, 'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'example_prompts': ['vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n', 'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\n', 'Compare and contrast artificial intelligence with human intelligence in terms of processing information.\n', 'Describe the basic components of a neural network and how it can be trained.\n', 'Write a short story about a robot that dreams for the first time.\n', 'Analyze the impact of the COVID-19 pandemic on global economic structures and future business models.\n', 'Explain the cultural significance of the Mona Lisa painting, and how its perception might vary in Western versus Eastern societies.\n', "Translate the following English sentence into Japanese, French, and Swahili: 'The early bird catches the worm.'\n"], 'model_name': 'PrunaAI/Einstein-v6.1-Llama3-8B-bnb-4bit-smashed', 'description': 'read pre-quantized 4-bit FP4 model'}
[2025-06-17T16:53:16Z] [31mFAILED[0m quantization/test_bitsandbytes.py::[1mtest_load_pre_quant_4bit_bnb_model[poedator/opt-125m-bnb-4bit-read pre-quantized 4-bit NF4 opt model][0m - AssertionError: function <function test_load_pre_quant_4bit_bnb_model at 0x7f4a42f18e00> failed when called with args () and kwargs {'hf_runner': <class 'tests.conftest.HfRunner'>, 'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'example_prompts': ['vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n', 'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\n', 'Compare and contrast artificial intelligence with human intelligence in terms of processing information.\n', 'Describe the basic components of a neural network and how it can be trained.\n', 'Write a short story about a robot that dreams for the first time.\n', 'Analyze the impact of the COVID-19 pandemic on global economic structures and future business models.\n', 'Explain the cultural significance of the Mona Lisa painting, and how its perception might vary in Western versus Eastern societies.\n', "Translate the following English sentence into Japanese, French, and Swahili: 'The early bird catches the worm.'\n"], 'model_name': 'poedator/opt-125m-bnb-4bit', 'description': 'read pre-quantized 4-bit NF4 opt model'}
[2025-06-17T16:53:16Z] [31mFAILED[0m quantization/test_bitsandbytes.py::[1mtest_load_8bit_bnb_model[meta-llama/Llama-Guard-3-8B-INT8-read pre-quantized llama 8-bit model][0m - AssertionError: function <function test_load_8bit_bnb_model at 0x7f4a42f19760> failed when called with args () and kwargs {'hf_runner': <class 'tests.conftest.HfRunner'>, 'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'example_prompts': ['vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n', 'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\n', 'Compare and contrast artificial intelligence with human intelligence in terms of processing information.\n', 'Describe the basic components of a neural network and how it can be trained.\n', 'Write a short story about a robot that dreams for the first time.\n', 'Analyze the impact of the COVID-19 pandemic on global economic structures and future business models.\n', 'Explain the cultural significance of the Mona Lisa painting, and how its perception might vary in Western versus Eastern societies.\n', "Translate the following English sentence into Japanese, French, and Swahili: 'The early bird catches the worm.'\n"], 'model_name': 'meta-llama/Llama-Guard-3-8B-INT8', 'description': 'read pre-quantized llama 8-bit model'}
[2025-06-17T16:53:16Z] [31mFAILED[0m quantization/test_bitsandbytes.py::[1mtest_load_8bit_bnb_model[yec019/fbopt-350m-8bit-read pre-quantized 8-bit opt model][0m - AssertionError: function <function test_load_8bit_bnb_model at 0x7f4a42f19760> failed when called with args () and kwargs {'hf_runner': <class 'tests.conftest.HfRunner'>, 'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'example_prompts': ['vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n', 'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\n', 'Compare and contrast artificial intelligence with human intelligence in terms of processing information.\n', 'Describe the basic components of a neural network and how it can be trained.\n', 'Write a short story about a robot that dreams for the first time.\n', 'Analyze the impact of the COVID-19 pandemic on global economic structures and future business models.\n', 'Explain the cultural significance of the Mona Lisa painting, and how its perception might vary in Western versus Eastern societies.\n', "Translate the following English sentence into Japanese, French, and Swahili: 'The early bird catches the worm.'\n"], 'model_name': 'yec019/fbopt-350m-8bit', 'description': 'read pre-quantized 8-bit opt model'}
[2025-06-17T16:53:16Z] [31mFAILED[0m quantization/test_bitsandbytes.py::[1mtest_4bit_bnb_embedding_model[half-intfloat/e5-mistral-7b-instruct-quantize embedding model inflight][0m - AssertionError: function <function test_4bit_bnb_embedding_model at 0x7f4a42f19c60> failed when called with args () and kwargs {'model_name': 'intfloat/e5-mistral-7b-instruct', 'description': 'quantize embedding model inflight', 'hf_runner': <class 'tests.conftest.HfRunner'>, 'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'example_prompts': ['vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.\n', 'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\n', 'Compare and contrast artificial intelligence with human intelligence in terms of processing information.\n', 'Describe the basic components of a neural network and how it can be trained.\n', 'Write a short story about a robot that dreams for the first time.\n', 'Analyze the impact of the COVID-19 pandemic on global economic structures and future business models.\n', 'Explain the cultural significance of the Mona Lisa painting, and how its perception might vary in Western versus Eastern societies.\n', "Translate the following English sentence into Japanese, French, and Swahili: 'The early bird catches the worm.'\n"], 'dtype': 'half'}
```

mgoin reviewed (approved):
Thanks for the back+forth and clean up, I think this looks ready to go now. I'll cc @bnellnm for additional review since he has been refactoring the MoE interface, but the `use_mxfp4_w4a4` addition should hopefully be not too problematic.

fxmarty-amd commented:
Hi @bnellnm, I'll fix conflicts and run tests again, I'll ping you again for review :)

fxmarty-amd committed:
Merge branch 'main' into mxfp4_moe

fxmarty-amd committed:
style

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
fix tests

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
linting

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

bnellnm commented on tests/quantization/test_quark.py:
Can you move any MoE specific parts of this test to a test file under tests/kernels/moe?  That way it will be easy to validate any MoE changes by running that one directory.

fxmarty-amd committed:
Merge branch 'main' into mxfp4_moe

fxmarty-amd committed:
fix updates with main and address comments

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
linting

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
linting 2

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd commented:
Hi @bnellnm, I addressed your comments and also made this compatible with the recent changes in vllm for dynamo/inductor, guarding mxfp4 dequantization & QDQ in custom ops.

Let me know if this looks good!

fxmarty-amd committed:
update doc

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
linting 3

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

mgoin reviewed (commented):
A few comments left

fxmarty-amd committed:
import fused_experts lazily

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd commented on vllm/model_executor/layers/quantization/quark/quark_moe.py:
```diff
@@ -214,8 +220,6 @@ def apply(
             raise NotImplementedError(
                 "EPLB not supported for `QuarkW8A8Fp8MoEMethod` yet.")
 
-        from vllm.model_executor.layers.fused_moe import fused_experts
```
fixed in https://github.com/vllm-project/vllm/pull/17888/commits/90a01bbd6371b5d5f83e77ac0ce948ee85b177ed

fxmarty-amd commented on vllm/model_executor/layers/quantization/quark/quark_moe.py:
```diff
@@ -242,4 +246,163 @@ def apply(
+            a1_scale=None,
+            a2_scale=None,
+            block_shape=None,
+            per_channel_quant=True,
```
Actually, this might not be compatible with https://github.com/vllm-project/vllm/blob/5358cce5ffbd4011f8fea2188995a249b43b8bfe/vllm/model_executor/layers/fused_moe/utils.py#L127-L144 which considers per-token quantization as having a single scale per token.

So I am not completely sure how to address https://github.com/vllm-project/vllm/pull/17888#discussion_r2172384925 - maybe it is okay for now not to add an assert check for per-token per-group?

fxmarty-amd committed:
pass activation arg

Signed-off-by: Felix Marty <Felix.Marty@amd.com>

fxmarty-amd committed:
remove per_channel_quant=True

Signed-off-by: Felix Marty <Felix.Marty@amd.com>
